### ISSUE TITLE
BottleneckCSP architecture

### ISSUE THREAD

**Issue by SISTMrL** [ultralytics/yolov5#781]
## ❔Question
I read the code bottleneckcsp in common.py, and it just convert the c channels into c/2 channels through convolution, the code is shown as follows:
`class BottleneckCSP(nn.Module):
    # CSP Bottleneck https://github.com/WongKinYiu/CrossStagePartialNetworks
    def __init__(self, c1, c2, n=1, shortcut=True, g=1, e=0.5):  # ch_in, ch_out, number, shortcut, groups, expansion
        super(BottleneckCSP, self).__init__()
        c_ = int(c2 * e)  # hidden channels
        self.cv1 = Conv(c1, c_, 1, 1)
        self.cv2 = nn.Conv2d(c1, c_, 1, 1, bias=False)
        self.cv3 = nn.Conv2d(c_, c_, 1, 1, bias=False)
        self.cv4 = Conv(2 * c_, c2, 1, 1)
        self.bn = nn.BatchNorm2d(2 * c_)  # applied to cat(cv2, cv3)
        self.act = nn.LeakyReLU(0.1, inplace=True)
        self.m = nn.Sequential(*[Bottleneck(c_, c_, shortcut, g, e=1.0) for _ in range(n)])

    def forward(self, x):
        y1 = self.cv3(self.m(self.cv1(x)))
        y2 = self.cv2(x)
        return self.cv4(self.act(self.bn(torch.cat((y1, y2), dim=1))))`
but I read the paper of CSPNet, in the figure, I see the CSPnet just separate the feature map into two part, I think the code above is different from the CSPnet, I couldn't understand
![1597759029(1)](https://user-images.githubusercontent.com/47938053/90522132-eff2c000-e19d-11ea-94d9-964a81e92280.jpg)
Could you tell me the reason, thanks very much!

**Comment by glenn-jocher** [comment]
@SISTMrL current implementation is correct.

**Comment by glenn-jocher** [comment]
@SISTMrL by the way, the best way to visualize the current models is probably with Netron viewing an pytorch model, or better an ONNX exported model.

**Comment by SISTMrL** [comment]
> @SISTMrL by the way, the best way to visualize the current models is probably with Netron viewing an pytorch model, or better an ONNX exported model.

ok, I'll have a try, thanks！

**Comment by abhiagwl4262** [comment]
@glenn-jocher Thanks for wonderful repo. I have few doubts regarding the CSPNet implementation. 

As per the paper the[https://openaccess.thecvf.com/content_CVPRW_2020/papers/w28/Wang_CSPNet_A_New_Backbone_That_Can_Enhance_Learning_Capability_of_CVPRW_2020_paper.pdf](url) , the input should be split into 2 parts which should be processed through 2 branches independently. But in your implementation, Both branches are taking the same input and without any spliting. 

Please correct me if I understood something wrong

**Comment by glenn-jocher** [comment]
@abhiagwl4262 yes that is correct. The inputs are not split, they are used in two places here, which I believe aligns with the actual CSPNet implementation.

**Comment by abhiagwl4262** [comment]
@glenn-jocher You guys have chosen to keep it this way because of mistake or for better performance based on experiment?

**Comment by glenn-jocher** [comment]
@abhiagwl4262 we have not tried the alternative methodology your asked about

**Comment by abhiagwl4262** [comment]
@glenn-jocher Ohhk. Thanks

**Comment by abhiagwl4262** [comment]
@glenn-jocher Hey, Is the input split suggested in CSPNet Paper is same as group convolution(Parallel Convolution) Where we are using groups=2? And What's your intuition behind the pros and cons of that ?

**Comment by glenn-jocher** [comment]
@abhiagwl4262 similar but not quite the same I think.  If I understand correctly the alternative interpretation of splitting the input channels may create small speed/size improvements at the cost of accuracy improvements, as is nearly always the case with reducing operations.

If in doubt you can update the CSP bottlenecks and train YOLOv5s to compare to the current implementation. The module is here:
https://github.com/ultralytics/yolov5/blob/702c4fa53eeaecfa5563c1441cb4a0c4aa8e908e/models/common.py#L49-L66

**Comment by marziyemahmoudifar** [comment]
What does BottleneckCSP do on the features extracted in the previous layers?

**Comment by glenn-jocher** [comment]
@marziyemahmoudifar `BottleneckCSP` operates on the extracted features from previous layers by applying a series of convolutions and transformations to the input, followed by a bottleneck structure that further refines the features. This process enhances the representational capacity of the network and facilitates the learning of more complex patterns in the feature maps. If you'd like to explore this further, you can refer to the documentation at https://docs.ultralytics.com/yolov5/.
